numpy/lib/recfunctions.py

```diff
@@ -870,6 +870,330 @@ def repack_fields(a, align=False, recurse=False):
     dt = np.dtype(fieldinfo, align=align)
     return np.dtype((a.type, dt))
 
+def _get_fields_and_offsets(dt, offset=0):
+    """
+    Returns a flat list of (name, dtype, count, offset) tuples of all the
+    scalar fields in the dtype "dt", including nested fields, in left
+    to right order.
+    """
+    fields = []
+    for name in dt.names:
+        field = dt.fields[name]
+        if field[0].names is None:
+            count = 1
+            for size in field[0].shape:
+                count *= size
+            fields.append((name, field[0], count, field[1] + offset))
+        else:
+            fields.extend(_get_fields_and_offsets(field[0], field[1] + offset))
+    return fields
+
+
+def _structured_to_unstructured_dispatcher(arr, dtype=None, copy=None,
+                                           casting=None):
+    return (arr,)
+
+@array_function_dispatch(_structured_to_unstructured_dispatcher)
+def structured_to_unstructured(arr, dtype=None, copy=False, casting='unsafe'):
+    """
+    Converts and n-D structured array into an (n+1)-D unstructured array.
+
+    The new array will have a new last dimension equal in size to the
+    number of field-elements of the input array. If not supplied, the output
+    datatype is determined from the numpy type promotion rules applied to all
+    the field datatypes.
+
+    Nested fields, as well as each element of any subarray fields, all count
+    as a single field-elements.
```

Comment:
What happens if `dtype` is itself a structured array? Eg, consider:

```
point = np.dtype([('x'. int), ('y', int)])
triangle = np.dtype([('a', point), ('b', point), ('c', point)]
```

I'd expect to be able to do
```
arr = np.zeros(10, triangle)

structured_to_unstructured(arr, dtype=point)
```

Reply:
This is accounted for, however your particular example shows there is a bug in this code because it can't account for repeated field names in the nested structures. Will fix.

Reply:
On second examination, I also missed that your output was unstructured. `structured_to_unstructured` doesn't account for that the way you expected, and I'm not sure there is a good "rule" for how it should work. Your example makes sense because each field can be unambiguously cast to the new structured type, so you expect the output shape to be (10, 3) with `point` dtype. My implementation currently casts all *nested* fields to the new dtype, resulting in a (10, 6) array of points: It casts the x and y of each point to a point individually.

Reply:
Yup, that's already fixed/implemented in #12446.

I'd rather not attempt to account for structured dtypes in the output though: That's not a pre-existing use-case we're trying to fix, and the best behavior to implement  is unclear to me at the moment. Any users who previously did something like that can still do it using `repack_fields` instead of `structured_to_unstructured`, though without the added safety the latter has added.

Reply:
> That's not a pre-existing use-case we're trying to fix,

As I understand it, the purpose of `structured_to_unstructured` is to replace the `arr[fields].view(dt)` idiom. But it sounds like it doesn't work as a universal replacement:

* `arr[['f_a', 'f_b']].view(float)`  &rarr; `structured_to_unstructured(arr['f_a', 'f_c'], float)`
* `arr[['p_a', 'p_b']].view(point)` &rarr; ???

Is there a way to spell the second case with these functions?

Reply:
Here's my understanding: 

All code of the form `arr[['field1', 'field2, ...]].view(dt)` in 1.15 can be replaced by `repack_fields(arr[['field1', 'field2, ...]]).view(dt)` in 1.16, without exceptions. It should be identical performance, since in both cases a copy is made.

Additionally, we have implemented a new function `structured_to_unstructured`. Although this can't be used as a replacement in all cases as you point out, in the many cases where it *can* be used it is better because it avoids a copy for multifield-indexes, it is safer since it is memory-layout-agnostic, and it better documents the user's intent. It is safer because it saves the user from bugs when doing the `view`: If the user tries to do the view themselves it is quite easy to forget to account for padding bytes, endianness, dtype, or misunderstand the memory layout, but `structured_to_unstructured` is written in a way to guarantee the view is "safe" for any memory layout and dtype (the fields are always viewed with the right offsets). `structured_to_unstructured` also casts the fields if needed, which the view above does not do.

Reply:
I added a description based on my last comment in #12447